src/useObserver.ts

```diff
@@ -24,42 +28,200 @@ export function useObserver<T>(
+                // BUT if we're a component that hasn't yet got to the useEffect()
+                // stage, we might be a component that _started_ to render, but
+                // got dropped, and we don't want to make state changes then.
+                // (It triggers warnings in StrictMode, for a start.)
+                if (trackingData.mounted) {
+                    // We have reached useEffect(), so we're mounted, and can trigger an update
+                    forceUpdate()
+                } else {
+                    // We haven't yet reached useEffect(), so we'll need to trigger a re-render
+                    // when (and if) useEffect() arrives.  The easiest way to do that is just to
+                    // drop our current reaction and allow useEffect() to recreate it.
+                    trackingData.reaction.dispose()
+                    reactionTrackingRef.current = null
+                }
+            })
+        }
+
+        reactionTrackingRef.current = trackingData
+        scheduleCleanupOfReactionIfLeaked(reactionTrackingRef)
     }
 
+    const reaction = reactionTrackingRef.current!.reaction
     useDebugValue(reaction, printDebugValue)
 
     useEffect(() => {
-        committed.current = true
-        return () => reaction.current!.dispose()
+        // Called on first mount only
+        recordReactionAsCommitted(reactionTrackingRef)
+
+        if (reactionTrackingRef.current) {
+            // Great. We've already got our reaction from our render;
+            // all we need to do is to record that it's now mounted,
+            // to allow future observable changes to trigger re-renders
+            reactionTrackingRef.current.mounted = true
+        } else {
+            // The reaction we set up in our render has been disposed.
+            // This is either due to bad timings of renderings, e.g. our
+            // component was paused for a _very_ long time, and our
+            // reaction got cleaned up, or we got a observable change
+            // between render and useEffect
+
+            // Re-create the reaction
+            reactionTrackingRef.current = {
+                reaction: new Reaction(observerComponentNameFor(baseComponentName), () => {
+                    // We've definitely already been mounted at this point
+                    forceUpdate()
+                }),
+                cleanAt: Infinity
+            }
+            forceUpdate()
+        }
+
+        return () => reactionTrackingRef.current!.reaction.dispose()
     }, [])
 
     // render the original component, but have the
     // reaction track the observables, so that rendering
     // can be invalidated (see above) once a dependency changes
     let rendering!: T
     let exception
-    reaction.current.track(() => {
+    reaction.track(() => {
         try {
             rendering = fn()
         } catch (e) {
             exception = e
         }
     })
     if (exception) {
-        reaction.current.dispose()
```

Comment:
Why was this disposal removed? If a component throws we can safely dispose (and also clean up the global state) since this very same instance is never going to come back?

Reply:
Regarding this, out of blue, I run into the issue when a reaction is disposed here and then component re-renders and the `track` fails to get executed on disposed reaction. I am not sure how is it possible, I tried to figure out some contained repro, but no luck. I suppose for the sake of such errors, the ref value should be unset here as well.

Reply:
If the component is going to be unmounted because of the error situation, it'll go through the cleanup logic in `useEffect`, which will already cause the reaction to be disposed. (Or, if it's throwing on first render, the cleanup work is already scheduled.)

I was attempting to reduce duplicated code as this particular `.dispose()` call is no longer needed.